### PR TITLE
feat: add GetNextPageCustomData to support request pages one by one

### PR DESCRIPTION
**Pull request by likyh**
### Summary
add GetNextPageCustomData to support request pages one by one.

### Does this close any open issues?
Related to #4292

### Screenshots

add test code in collector:
```golang
	collector, err := helper.NewApiCollector(helper.ApiCollectorArgs{
		……
		PageSize:           50,
		Query: func(reqData *helper.RequestData) (url.Values, errors.Error) {
			query := url.Values{}
			query.Set("state", "all")
			query.Set("pagelen", fmt.Sprintf("%v", reqData.Pager.Size))
			if reqData.CustomData != nil {
				query.Set("page", reqData.CustomData.(string))
			}
			return query, nil
		},
		GetNextPageCustomData: func(reqData *helper.RequestData, prevPageResponse *http.Response) (interface{}, errors.Error) {
			resBody, err := io.ReadAll(prevPageResponse.Body)
			// decode json from body
			if rawMessages.Next == `` {
				return ``, helper.ErrFinishCollect
			}
			u, err := url.Parse(rawMessages.Next)
			println(u.Query()[`page`][0])
			return u.Query()[`page`][0], nil
		},
	})
```

api response:
image

log:
```
time="2023-02-03 18:09:51" level=info msg=" [collectApiPullRequestCommits] start api collection"
time="2023-02-03 18:09:52" level=info msg=" [collectApiPullRequestCommits] finished records: 1"

# the first number means the page
# the second json means which pr is
# the third number means what the next page token is
1 {"BitbucketId":1} xGCox
1 {"BitbucketId":3}
1 {"BitbucketId":2} xGCox
2 {"BitbucketId":1} 7nSyG
2 {"BitbucketId":2} 7nSyG
3 {"BitbucketId":1} LRCkr
3 {"BitbucketId":2} LRCkr
4 {"BitbucketId":1} RaSbR
4 {"BitbucketId":2} RaSbR
time="2023-02-03 18:09:55" level=info msg=" [collectApiPullRequestCommits] finished records: 10"
5 {"BitbucketId":1}
5 {"BitbucketId":2}
time="2023-02-03 18:09:56" level=info msg=" [collectApiPullRequestCommits] end api collection without error"
time="2023-02-03 18:09:56" level=info msg="finished step: 5 / 5"
```

We can find 3 pr's commits are collected parallelly and each page in a PR are collected in order.

### CODE REVIEW

File: backend/helpers/pluginhelper/api/api_collector.go [hunk 1]
```diff
@@ -208,13 +212,43 @@ func (collector *ApiCollector) exec(input interface{}) {
...
+		collector.fetchPagesInOrder(reqData)
 	} else if collector.args.GetTotalPages != nil {
 		collector.fetchPagesDetermined(reqData)
 	} else {
 		collector.fetchPagesUndetermined(reqData)
 	}
 }
 
+// fetchPagesInOrder fetches data of all pages in order to build query by prev response
+func (collector *ApiCollector) fetchPagesInOrder(reqData *RequestData) {
```
Comment (klesh) [inline backend/helpers/pluginhelper/api/api_collector.go:225]: How about `fetchPagesSequentially`?

File: backend/helpers/pluginhelper/api/api_collector.go [hunk 2]
```diff
@@ -208,13 +212,43 @@ func (collector *ApiCollector) exec(input interface{}) {
 	}
 	if collector.args.PageSize <= 0 {
 		collector.fetchAsync(reqData, nil)
+	} else if collector.args.PageInOrder {
```
Comment (klesh) [inline backend/helpers/pluginhelper/api/api_collector.go:215]: I think we may define a `GetNextPagerUrl` function on par with `GetTotalPages`

File: backend/helpers/pluginhelper/api/api_collector.go (hunk 1 above)
Comment (likyh) [reply backend/helpers/pluginhelper/api/api_collector.go]: fixed

File: backend/helpers/pluginhelper/api/api_collector.go (hunk 2 above)
Comment (likyh) [reply backend/helpers/pluginhelper/api/api_collector.go]: add `GetNextPageCustomData` on par with `GetTotalPages`

**Review by klesh — APPROVED** [review]
LGTM

**Pull request merged**